### PR TITLE
LSC2-246: updates to synch with DBUP

### PR DESCRIPTION
**Pull request by marwanrahman**
## What issues/features does this PR address?
Updating the repo to function with latest DBUP code. 

## Description of the new behaviour/code
Mostly fixing outdated code and adding new params

### Manual tests:
Can run dataset creator successfully now.

## Other notes/questions for reviewers
This does not address querying from a csv, that change will come after.

### CODE REVIEW

**Review by delaneysmith22 — APPROVED** [review]
Great work, Marwan! My main comment would be about the `video_to_mmodes ` function - I think needs to be re-worked a little bit. I know this wasn't the purpose of this PR, but it may be a good time as we move towards the version of the repo that we will be using for re-training.

File: src/data/dataset_creator.py [hunk 1]
```diff
@@ -104,166 +204,307 @@ def build_dataset(self) -> None:
...
         if self.external:
-            miniclip_csv_path = self.cfg['EXTERNAL_VAL']['PATHS']['MINI_CLIPS_TABLE'].replace('LOCATION', self.external)
+            miniclip_csv_path = self.cfg['EXTERNAL_VAL']['PATHS']['MINI_CLIPS_TABLE'].replace(
+                'LOCATION', self.external)
         else:
             miniclip_csv_path = self.cfg['PREPROCESS']['PATHS']['MINI_CLIPS_TABLE']
-        
-        miniclip_df.to_csv(miniclip_csv_path, index=False)
 
+        miniclip_df.to_csv(miniclip_csv_path, index=False)
```
[the remark this answers is not in the document]
Comment (marwanrahman) [reply src/data/dataset_creator.py:341]: You're right, I added those in there

File: src/data/dataset_creator.py [hunk 2]
```diff
@@ -104,166 +204,307 @@ def build_dataset(self) -> None:
...
-            if file[-4:] == (".csv"):
+            if file[-4:] == ".csv" and (
+                    file != os.path.basename(clips_to_query_csv_path)
+            ):
                 os.remove(os.path.join(csv_output_path, file))
 
         # Get database connection via ssh tunnel
-        server, conn = self.get_db_cnx()
-        
+        conn, server = self.get_db_cnx()
```
[the remark this answers is not in the document]
Comment (marwanrahman) [reply src/data/dataset_creator.py:372]: We're still connecting either way to get masked clips and m mode rows based on the provided csv

File: src/data/dataset_creator.py [hunk 3]
```diff
@@ -104,166 +183,342 @@ def build_dataset(self) -> None:
...
-        miniclip_df = pd.DataFrame(miniclip_df_rows, columns=['id', 'miniclip_id', 'patient_id'])
+        miniclip_df = pd.DataFrame(
+            miniclip_df_rows,
+            columns=['id', 'miniclip_id', 'patient_id']
+        )
+
+        mmode_df = pd.DataFrame(
+            mmode_df_rows, columns=[
+                'id', 'miniclip_id', 'pixel_x_index', 'sum_brightness',
+                'Frame Path', 'brightness_rank',
```
Comment (bvanberl) [inline src/data/dataset_creator.py:242]: Very picky, but would it be worthwhile changing `Frame Path` to snake case to homogenize the column names?

File: src/data/dataset_creator.py [hunk 4]
```diff
@@ -104,166 +183,342 @@ def build_dataset(self) -> None:
...
+                by=['miniclip_id', 'sum_brightness'],
+                inplace=True,
+                ascending=False
+            )
+            df_to_pull.loc[:, 'brightness_rank'] = df_to_pull.groupby(
+                'miniclip_id')['sum_brightness'].rank(ascending=False)
+
+            # filter m modes based on rank
+            num_mmodes = self.cfg['PREPROCESS']['PARAMS']['NUM_MMODES']
+            df_to_pull = df_to_pull.groupby('miniclip_id').head(num_mmodes)
```
Comment (bvanberl) [inline src/data/dataset_creator.py:274]: Wondering if we should give the option to filter based on top % as well as top #.

File: src/data/dataset_creator.py [hunk 5]
```diff
@@ -104,166 +183,342 @@ def build_dataset(self) -> None:
...
+
+            # pull preprocessed m modes
+            self.database_pull(
+                df_to_pull,
+                path=png_folder,
+                blob_key='m_mode_resized_blob_path',
+            )
+
+            # append new rows to mincilip dataframe
```
Comment (bvanberl) [inline src/data/dataset_creator.py:283]:
```suggestion
            # append new rows to miniclip dataframe
```

File: src/data/dataset_creator.py [hunk 6]
```diff
@@ -104,166 +204,307 @@ def build_dataset(self) -> None:
...
-
-        # Merge with bbox df and mmode df
-        miniclip_df = pd.merge(miniclip_df, bboxes_df, 'inner', on=['id', 'miniclip_id'])
-        miniclip_df = pd.merge(mmode_df, miniclip_df, 'left', on=['id', 'miniclip_id'])
-        miniclip_df['miniclip_id'] = miniclip_df['id'].astype(str) + "_" \
-                                        + miniclip_df['miniclip_id'].astype(str)
-        miniclip_df['mmode_id'] = miniclip_df['miniclip_id'].astype(str) + "_" \
-                                            + miniclip_df['Pixel Index'].astype(str)
-
+            mmode_df = mmode_df.drop(['brightness_rank'], axis=1)
```
[the remark this answers is not in the document]
Comment (bvanberl) [reply src/data/dataset_creator.py:300]: Agreed.

File: src/data/dataset_creator.py [hunk 7]
```diff
@@ -104,166 +183,342 @@ def build_dataset(self) -> None:
...
+                'CSVS_OUTPUT'].replace('LOCATION', self.external)
+            clips_to_query_csv_path = self.cfg['EXTERNAL_VAL']['PATHS'][
+                'CLIPS_TO_QUERY']
         else:
             csv_output_path = self.cfg['PREPROCESS']['PATHS']['CSVS_OUTPUT']
+            clips_to_query_csv_path = self.cfg['PREPROCESS']['PATHS'][
+                'CLIPS_TO_QUERY']
         for file in os.listdir(csv_output_path):
-            if file[-4:] == (".csv"):
+            if file[-4:] == ".csv" and (
```
Comment (bvanberl) [inline src/data/dataset_creator.py:366]: It might be safer to use `os.path.splitext(file)[1]`

File: src/data/dataset_creator.py [hunk 8]
```diff
@@ -104,166 +183,342 @@ def build_dataset(self) -> None:
...
             query_table_path = self.cfg['PREPROCESS']['PATHS']['QUERY_TABLE']
         print('Writing to ' + query_table_path + "...")
         clips_df.to_csv(query_table_path, index=False)
 
         # Finalize dataframe
-        clips_df = clips_df[COLUMNS_WANTED + ['Class']]
+        clips_df = assign_subclass(clips_df)
+        clips_df = clips_df[columns_wanted]
+
+        # TEMPORARY - for excluding calibration set due to missing
```
Comment (bvanberl) [inline src/data/dataset_creator.py:458]: I'd add a `TODO` here just to make sure this is addressed later on.

File: src/data/dataset_creator.py [hunk 9]
```diff
@@ -303,16 +559,16 @@ def assign_subclass(df: pd.DataFrame):
...
-        'thickened|lung_point': 'lung_point',
-        'absent_lung_sliding|lung_point': 'lung_point',
-        'lung_pulse': 'lung_pulse',
-        'absent_lung_sliding|lung_pulse': 'lung_pulse',
-        'thickened|lung_pulse': 'lung_pulse',
+        'Absent lung sliding': 'normal_absent_lung_sliding',
+        'Thickened': 'lung_sliding',
+        'Lung point': 'lung_point',
+        'Thickened|Absent lung sliding': 'normal_absent_lung_sliding',
+        'Absent lung sliding|Thickened': 'normal_absent_lung_sliding',
```
Comment (bvanberl) [inline src/data/dataset_creator.py:566]: "normal_absent_lung_sliding" is clinically misleading in my opinion. Maybe replace with "regular_absent_lung_sliding"?

File: src/data/dataset_creator.py [hunk 10]
```diff
@@ -386,41 +643,43 @@ def balance_train(train_df: pd.DataFrame) -> pd.DataFrame:
...
+        try:
+            # Currently, memory growth needs to be the same across GPUs
+            for gpu in gpus:
+                tf.config.experimental.set_memory_growth(gpu, True)
+            logical_gpus = tf.config.list_logical_devices('GPU')
+            print(len(gpus), "Physical GPUs,", len(logical_gpus),
+                  "Logical GPUs")
+        except RuntimeError as e:
+            # Memory growth must be set before GPUs have been initialized
+            print(e)
```
Comment (bvanberl) [inline src/data/dataset_creator.py:667]: Not sure if we need this anymore

File: src/data/to_png.py [hunk 11]
```diff
@@ -227,89 +101,152 @@ def video_to_mmodes(path: str,
                 # deterministic method should not be given resized image
                 bbox_pred_frame = copy.deepcopy(frame)
 
-            weights = [0.2989, 0.5870, 0.1140]  # In accordance with tfa rgb to grayscale
+            weights = [0.2989, 0.5870,
+                       0.1140]  # In accordance with tfa rgb to grayscale
             grayscale_frame = np.dot(grayscale_frame, weights).astype(np.uint8)
             grayscale_frame = np.expand_dims(grayscale_frame, axis=-1)
```
Comment (bvanberl) [inline src/data/to_png.py:107]: I've wanted to change this for a while -- why don't we just use cv2.cvtColor and convert from colour (check if BGR or RGB) to greyscale?

File: src/external_validation.py [hunk 12]
```diff
@@ -120,15 +120,17 @@ def evaluate_model(model : tf.keras.Model,
     :return: dict of evaluation metrics
     '''
 
-    test_set = tf.data.Dataset.from_tensor_slices((test_df['Frame Path'].tolist(), test_df['Class'].tolist()))
+    test_set = tf.data.Dataset.from_tensor_slices(
+        (test_df['Frame Path'].tolist(), test_df['Class'].tolist()))
```
Comment (bvanberl) [inline src/external_validation.py:124]: If we change `Frame Path`, will have to search across the whole repo for places to replace it

File: src/external_validation.py [hunk 13]
```diff
@@ -213,7 +215,9 @@ def minority_upsample(df : pd.DataFrame,
     '''
     counts = df['Class'].value_counts()
     n_additional = math.ceil((1. / (1. - minority_frac) - 1) * counts.max()) - counts.min()
-    absent_train_df = df.loc[df['Class'] == 1].copy().sample(frac=1, random_state=seed)
+    absent_train_df = df.loc[
+        df['Class'] == 1
```
Comment (bvanberl) [inline src/external_validation.py:219]: Random thought -- should we make an [enum](https://www.geeksforgeeks.org/enum-in-python/),  or globally defined constants, for the classes? It might help to ensure the issue of the class flipflop is canonically settled forever.

e.g.
```
class LungSlidingClass(Enum)
    PRESENT = 0
    ABSENT = 1
```

Then it could be accessed everywhere with, for example, `LungSlidingClass.ABSENT`
Naming could be different, maybe more condensed.

**Review by bvanberl — APPROVED** [review]
Superb work @marwanrahman !!

File: src/external_validation.py (hunk 13 above)
Comment (marwanrahman) [reply src/external_validation.py:219]: Totally agree, much cleaner and safer that way

File: src/preprocessor.py
Comment (bvanber) [inline src/preprocessor.py:1]: Should any of these functions live in deep breathe utils?

**Review by bvanber — APPROVED** [review]
